src/mem3/src/mem3_util.erl

```diff
@@ -508,19 +512,27 @@ sort_ranges_fun({B1, _}, {B2, _}) ->
     B1 =< B2.
 
 
-get_or_create_db(DbName, Options) ->
-    case couch_db:open_int(DbName, Options) of
+add_db_config_options(DbName, Options) ->
+    DbOpts = case mem3:dbname(DbName) of
+        DbName  -> [];
+        MDbName -> mem3_shards:opts_for_db(MDbName)
+    end,
+    merge_opts(DbOpts, Options).
+
+
+get_or_create_db(DbName, Options0) ->
+    Options = case proplists:get_value(user_ctx, Options0, undefined) of
```

Comment:
I noticed that we switched from using `couch_db:open_int/2` to `couch_db:open/2`. The former doesn't throw any reader errors, and the latter checks for that the `user_ctx` is valid. Is there a particular reason you are making this change. I presume it has to deal with being more restrictive?

Reply:
Yeah... I should have mentioned what I did here, although it's a bit more detailed in the commits getting to this stage. Basically we originally had two codepaths to account for, `couch_server:open with create_if_missing=true` and direct calls to `couch_server:create`. It sure looks to me that there's no relevant difference between using either code path, as they both end up going through the same `couch_server:open_async` path and both return valid database handles. So rather than abstracting over whether to use `couch_server:create` or `couch_server:open`, I switched to only using the `open` variant and having everything funnel through that.

For the most part, this works just fine, as code that gets here usually supplies a `user_ctx` value, or embeds `?ADMIN_CTX`, for instance here's one of the places we do this in `fabric:get_security` [1], or similarly [2][3]. Basically these code paths count on having a `user_ctx` set, or embedding `?ADMIN_CTX` as the `user_ctx`. So I switched to checking for a `user_ctx` here, and embedding `?ADMIN_CTX` if it's missing, and this approach allowed for all of the different code paths to use the same core db creation logic, and it has the added benefit of not needing to preload the db config properties until a missing shard is encountered, avoiding doing a dedicated doc lookup on the dbs db for any of the `fabric_util:get_db` cases (like fabric:get_security) which is called often.

To answer your direct question, the motivation of switching to `couch_db:open` is to use the normal open logic that validates the `user_ctx` is allowed to access the database, allowing us to use this function for normal user initiated operations where there's a valid user_ctx, along with internal situations where shards need to be created after the fact.


[1] https://github.com/apache/couchdb/blob/3.x/src/fabric/src/fabric.erl#L177-L183
[2] https://github.com/apache/couchdb/blob/3.x/src/fabric/src/fabric.erl#L174
[3] https://github.com/apache/couchdb/blob/3.x/src/fabric/src/fabric.erl#L152

Reply:
There may be a regex execution for db name validation in https://github.com/apache/couchdb/blob/d2c123b4096e39b6fcd1f9ac19c061bd2fcd6cec/src/couch/src/couch_server.erl#L733 which is called from `open/2` but not from `open_int/2` if  I read that code correctly.

The performance impact may not be too bad (10 - 100 microseconds?) for one run, but could it add up when multiple shards are open, especially with this being called for every internal shard open (in `fabric_rpc:with_db/3` calls, all internal replicator opens in `mem3_rpc`). 

Also wonder if db name validation rules are ever changed, or if there are invalid db names in an existing cluster, if this would break internal replicator and prevent it from making any progress, while previously it might have broken on user requests only.

Reply:
Good catch on `open` calling `validate_open_or_create` and `open_int` not doing so. I noticed it's being done in `couch_server:create` so I wasn't terribly worried about it happening here. `dreyfus_rpc` is using `couch_server:create` (as is `hastings_rpc`) so we've already got scenarios where shard logic may be applied later.

In terms of performance, my biggest concern is around the time associated with loading the underlying dbs db doc, as it's not cached and if we have to preload it everytime on the coordinator side, then that'll add a lot of overhead to most rpc requests. My goal with the most recent restructuring of this code was to minimize the instances where we actually need to load the dbs db doc, and wait until the last possible moment to do it.

Reply:
Makes sense regarding loading the dbs db doc. I was mainly wondering about running regex validation code on every single shard open in [fabric_rpc](https://github.com/apache/couchdb/blob/3.x/src/fabric/src/fabric_rpc.erl#L325) and [mem3_rpc](https://github.com/apache/couchdb/blob/3.x/src/mem3/src/mem3_rpc.erl#L403). We didn't do it before and not sure if we can avoid the slight performance hit and if it might alter behavior on clusters which somehow already have invalid db names. May be we could have a `mem3_util:get_or_create_int/2` or add an `{internal, true}` flag?